### PR TITLE
QSPI tests

### PR DESCRIPTION
**Pull request by bjoernQ**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [ ] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description
This is based on #2013 - so merge that first.

This adds basic tests for QSPI

#### Testing
Tests works and they fail e.g. when reverting changes from #2013

### CODE REVIEW

File: hil-test/tests/qspi_read.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        x
+    }};
+}
+
+struct Context {
+    spi: esp_hal::peripherals::SPI2,
+    dma_channel: Channel<'static, DmaChannel0, Blocking>,
+    miso: esp_hal::gpio::GpioPin<2>,
+    miso_mirror: Output<'static, GpioPin<3>>,
+    clocks: &'static Clocks<'static>,
```
Comment (bugadani) [inline hil-test/tests/qspi_read.rs:58]: Can this just be `Clocks<'static>`?

File: hil-test/tests/qspi_write.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+//! This uses PCNT to count the edges of the MOSI signal
+//!
+//! Following pins are used:
+//! MOSI    GPIO2
+//!
+//! PCNT    GPIO3
+//!
+//! Connect MOSI (GPIO2) and PCNT (GPIO3) pins.
+
+//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s2 esp32s3
```
Comment (bugadani) [inline hil-test/tests/qspi_write.rs:12]: I think we are limited to `//% CHIPS: esp32 esp32c6 esp32h2 esp32s3` with PCNT

**Comment by Dominaezzz** [comment]
Mind adding a test that alternates between writing and reading? I'm worried https://github.com/esp-rs/esp-hal/issues/1860 might also affect half duplex as well.

**Review by jessebraham — APPROVED** [review]
Thank you for this!

**Review by SergioGasquez — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**